File: stacks_and_queues/questions/question_3.5/problem_3.5.rb
class Node
  attr_accessor :value, :next

  def initialize value=nil
    @value = value
    @next = nil
  end
end

class Stack

  attr_accessor :top

  def initialize top=nil
    @top = Node.new top
  end

  def pop
    return nil unless top
    pop_result = top.value
    @top = top.next
    pop_result
  end

  def push item
    item = Node.new(item)
    item.next = top
    @top = item
  end

  def peek
    top.value
  end
end

class MyQueue

  def initialize
    @main_stack     = Stack.new
    @transfer_stack = Stack.new
  end

  def enqueue value
    while @main_stack.peek
      @transfer_stack.push(@main_stack.pop)
    end
    @transfer_stack.push value
    while @transfer_stack.peek
      @main_stack.push(@transfer_stack.pop)
    end
  end

  def dequeue
    @main_stack.pop
  end

  def peek
    @main_stack.peek
  end
end


mq = MyQueue.new
[1,2,3,4,5].each { |val| mq.enqueue val }
while mq.peek
  p mq.dequeue
end
p mq.dequeue

File: stacks_and_queues/questions/question_3.5/problem_3.5.js
var Node = function(value){
  timer = Date.now()
  while (Date.now() === timer){}
  this.value = value
  this.id = Date.now()
  this.next = null
};

var Stack = function(value){
  if (value) {this.top = new Node(value)}
};

Stack.prototype.pop = function(){
  if (!this.top) {return null}
  result = this.top
  this.top = this.top.next
  return result.value
}

Stack.prototype.push = function(item){
  nodeItem = new Node(item)
  nodeItem.next = this.top
  this.top = nodeItem
}

Stack.prototype.peek = function(){
  return (this.top ? this.top.value : null)
}

var MyQueue = function(){
  this.mainStack = new Stack()
  this.transferStack = new Stack()
}

MyQueue.prototype.enqueue = function(value){
  while(this.mainStack.peek()){
    this.transferStack.push(this.mainStack.pop())
  }
  this.transferStack.push(value)
  while(this.transferStack.peek()){
    this.mainStack.push(this.transferStack.pop())
  }
}

MyQueue.prototype.dequeue = function(){
  return this.mainStack.pop()
}

MyQueue.prototype.peek = function(){
  return this.mainStack.peek()
}


var arr = [1,2,3,4,5]
var mQ = new MyQueue()
for(var val of arr){
  mQ.enqueue(val)
}
while(mQ.peek()){
  console.log(mQ.dequeue())
}
console.log(mQ.dequeue())